src/main/java/algorithms/curated170/medium/WordPattern2.java

```diff
@@ -0,0 +1,111 @@
+package algorithms.curated170.medium;
+
+import java.util.HashSet;
+
+public class WordPattern2 {
+
+    String[] mapMatches;
+    HashSet<String> setMatches;
+    String pattern, str;
+
+    public boolean wordPatternMatch(String pattern, String str) {
+        mapMatches = new String[26]; 
+        setMatches = new HashSet<>(); 
+        this.pattern = pattern;
+        this.str = str;
+        return backtrackPatternMatch(0, str.length() - 1, 0, pattern.length() - 1);
+    }
+
+    private boolean backtrackPatternMatch(int start, int end, int pStart, int pEnd) {
+
+        if (pStart == pEnd + 1 && start == end + 1) {
+            return true;
+        } else if ((pStart > pEnd) || (start > end)) {
+            return false;
+        }
+
+        char ch = pattern.charAt(pStart);
+        int chIdx = letterIndex(ch);
+        String matched = mapMatches[chIdx];
+
+        if (matched != null) {
+            return isValidMatch(start, end, pStart, pEnd, matched);
+        }
+
+        return isPossibleToMatch(start, end, pStart, pEnd, chIdx);
+    }
+
+    private boolean isPossibleToMatch(int start, int end, int pStart, int pEnd, int chIdx) {
+        
+        int unmatchedEndPos = findLastUnmatchedPos(end, pStart, pEnd);
+        
+        for (int i = start; i <= unmatchedEndPos; i++) {
+
+            String matched = str.substring(start, i + 1);
+
+            if (setMatches.contains(matched)) {
+                continue;
+            }
+
+            mapMatches[chIdx] = matched;
+            setMatches.add(matched);
+
+            if (backtrackPatternMatch(i + 1, end, pStart + 1, pEnd)) {
+                return true;
+            } else {
+                cleanPossibleMatch(chIdx, matched);
+            }
+        }
+
+        return false;
+    }
+
+    private int findLastUnmatchedPos(int end, int pStart, int pEnd) {
+        int unmatchedEndPos = end;
+
+        for (int i = pEnd; i > pStart; i--) {
+            unmatchedEndPos -= patternsToMatch(i);
+        }
+        return unmatchedEndPos;
+    }
+
+    private int patternsToMatch(int i) {
+        return mapMatches[letterIndex(pattern.charAt(i))] == null ? 1
+                : mapMatches[letterIndex(pattern.charAt(i))].length();
+    }
```

Comment:
```suggestion

```

I am not sure if it is any better than:
    
`str.length() - (pattern.length() - pStart)`

Reply:
If we have "abc"  at the end of the string, each is perhaps assigned to patterns length 10. We would then have to subtract  a total of 30 length from the end position.

Reply:
There is this constraint: `1 <= pattern.length, s.length <= 20`

Reply:
Some patterns are already matched to substrings within the string. So assuming that these already matched patterns are of length 1 is not right.